docs/report_stage_i.md

```diff
@@ -0,0 +1,497 @@
+
+#### Simplest valid program with simple surroundings
+
+Before a full-fledged support for exploring various surroundings and their impact on the execution of a given OPCODE, we can provide a simplified probe of the topic.
+
+In this approach, we will generate programs same as in "Simplest valid program" approach, but:
+  - prepend and/or append a sequence of fixed arbitrary bytecodes, to put the analyzed OPCODE in the "middle" of the program, and/or
+  - multiply the OPCODE with its stack management companions.
+
+Such programs can then be used for initial exploration of the impact of surroundings.
+
+#### Looped execution with stack balancing
+
+See [Benchmarking EVM Instructions by @chfast](https://notes.ethereum.org/@chfast/benchmarking-evm-instructions) (**TODO ref**).
+
+This approach will be very useful to validate the individual instruction measurements with whole-program measurements.
+
+#### Completely randomized with stack balancing
+
+This is an extension of the "Simplest valid program" approach, where we'll generate random programs, mixing instructions of all OPCODEs from the subset, that only are known to execute successfully due to correct stack balancing.
+
+This approach will be very useful to validate the individual instruction measurements with whole-program measurements ("measure total").
+Also it will be applicable as the first step towards exploring the sources of variability of OPCODE cost driven by various contexts in which these OPCODEs are executed.
+
+#### Automated, adaptive generation
+
+This is somewhat similar to the approach from the ["Broken metre" paper](https://arxiv.org/pdf/1909.07220.pdf) (**TODO make ref**), which describes a genetic algorithm minimizing the gas throughput (gas/second).
+Pursuing this possibility is an optional expansion to the scope of the research, which will be considered at Stage II.
+
+We could modify this idea to run a genetic algorithm (or any other adaptive method with an objective function) maximizing our desired properties, i.e.:
+  - maximize variance (information) captured in the measurement of a particular OPCODE, in particular circumstances (particular `program`, particular location)
+  - minimize variance (noise) coming from different environments
+  - maximize uniformity
+
+This could also be useful to discover the impact of circumstances ("surrounding") (**TODO unify language circumstances, surrounding, context**), where a particular OPCODE can exhibit much different computational burden, depending on where it is located in the program.
+
+One challenge is how to model OPCODEs with variable size of parameters (i.e. `SHA3`): they would have unbounded values and also unbounded variance, if not modeled correctly.
+We could model as `t = c + a*x`, `c` constant, `x` size, `a` coefficient to model.
+Then, instead of most variance/information about `t`, we want most information on `c` and `a`.
```

Comment:
Perhaps it's worth noting here that the current cost of SHA3 is , IIRC, 30+6*x

Reply:
@mbenke, note that here we're talking about modeling time/computational cost variability, not current gas schedule. I'll add this remark in one of the earlier mentions of `SHA3`